### PR TITLE
[WIP] Add img2img

### PR DESCRIPTION
**Pull request by ayushtues**
Adding img2img pipeline for Kandinksy, part of #3308

### CODE REVIEW

**Comment by ayushtues** [comment]
@yiyixuxu the UnCLIP scheduler doesn't have an `add_noise` function, which is needed for img2img to noise the image to a particular timestep, I am not familiar with UnCLIP scheduler, but is it the same as DDPM?

**Comment by yiyixuxu** [comment]
>  I am not familiar with UnCLIP scheduler, but is it the same as DDPM?

yes yes!! see this comments here https://github.com/huggingface/diffusers/pull/3308#issuecomment-1545908608 - let's just use DDPM for img2img. We will need to swap unclip scheduler with DDPM for other 2 pipelines too! but we can wait to do that later

**Comment by ayushtues** [comment]
Got some initial results working 

```python
from diffusers import KandinskyPipeline, KandinskyPriorPipeline, KandinskyImg2ImgPipeline
# from diffusers.src.pipelines import KandinskyImg2ImgPipeline
from transformers import AutoTokenizer
import requests
from PIL import Image
from io import BytesIO
from diffusers import DDPMScheduler
import torch
import numpy as np
import gc

ddpm_config = {
  "clip_sample": True,
  "clip_sample_range": 2.0,
  "sample_max_value": None,
  "num_train_timesteps": 1000,
  "prediction_type": "epsilon",
  "variance_type": "learned_range",
  "thresholding": True,
  "beta_schedule": "linear",
  "beta_start": 0.00085,
  "beta_end":0.012
}

url = "https://preview.redd.it/yu4maxz3dxo91.jpg?width=1024&format=pjpg&auto=webp&v=enabled&s=64ebd870f1f0cab6c94b5ee75ca03a53a1070068"
response = requests.get(url)
init_image = Image.open(BytesIO(response.content)).convert("RGB")
init_image = init_image.resize((768, 768))

prompt = "A cartoon frog"
device = 'cuda'
batch_size=1 


# # create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained("YiYiXu/Kandinsky-prior")
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
image_emb = pipe_prior(prompt, generator=generator,)
zero_image_emb = pipe_prior("")


pipe = KandinskyPipeline.from_pretrained("YiYiXu/Kandinsky")
ddpm = DDPMScheduler(**ddpm_config)
generator = torch.Generator(device=device).manual_seed(0)

pipe_img2img = KandinskyImg2ImgPipeline(text_encoder=pipe.text_encoder, tokenizer=pipe.tokenizer, text_proj=pipe.text_proj, unet=pipe.unet, scheduler=ddpm, movq=pipe.movq)
pipe_img2img.to(device)
out = pipe_inpainting(prompt=prompt, image=init_image, height=768, width=768, num_inference_steps=100, generator=generator, image_embeds=image_emb, negative_image_embeds=zero_image_emb, strength=0.2) 

out[0][0]

```
Before img2img
image


After img2img
image

**Comment by yiyixuxu** [comment]
@ayushtues 
awesome! Did you compare it with the results from the original repo?

**Comment by yiyixuxu** [comment]
@ayushtues 

I think we will have to change the code slightly to make it match entirely - I did that for text2img. I can help run it to make sure once this is merged in 
maybe you can try running the original repo with `p_sampler`?  the https://github.com/ai-forever/Kandinsky-2/blob/main/kandinsky2/kandinsky2_1_model.py#L438 

anyways let me know when this is ready to merge in - we can always make changes later from the base PR! 

and also do you have a twitter handle?

**Comment by yiyixuxu** [comment]
@ayushtues 

ohh another thing you should try is to replace the Unclip scheduler in the text2img pipeline to use DDPM  - that should be easy since you already did for img2img! And we will need to see the same results as unclip scheduler for that

**Comment by ayushtues** [comment]
@yiyixuxu I replaced UnCLIP in text2img, and found that it produced different images, that's because there seem to be slight differences between UnCLIP and DDPM, namely

1. The `step_ratio` calculation - [DDPM](https://github.com/huggingface/diffusers/blob/793e408fc686915c35aa563fd42aac8274f1f60f/src/diffusers/schedulers/scheduling_ddpm.py#L232) vs [UnCLIP](https://github.com/huggingface/diffusers/blob/793e408fc686915c35aa563fd42aac8274f1f60f/src/diffusers/schedulers/scheduling_unclip.py#L176)
2. Thresholding and Clipping - Its either threshold _or_ clipping in [DDPM](https://github.com/huggingface/diffusers/blob/793e408fc686915c35aa563fd42aac8274f1f60f/src/diffusers/schedulers/scheduling_ddpm.py#L370) but its _both_ threshold and clipping in [UnCLIP](https://github.com/huggingface/diffusers/blob/793e408fc686915c35aa563fd42aac8274f1f60f/src/diffusers/schedulers/scheduling_unclip.py#L320) 

So getting the same results from them is not possible, since they are different in implementation

**Comment by ayushtues** [comment]
Other than that, I think we can merge the repo, and then make changes to the schedulers in the base PR itself

**Comment by yiyixuxu** [comment]
@ayushtues I see - I will update our DDPM scheduler to make sure it works for our model. Don't worry about that for now

Can you add a test for img2img like what I did here for inpainting? https://github.com/huggingface/diffusers/pull/3308/files#diff-a94251ed6c7af41b0a066c4bc7cc78cadc0e0fe9c766c51ed5d308a383626cc1

**Comment by yiyixuxu** [comment]
@ayushtues I merged in and will make changes from my PR -  Thanks for the great work!
